iroh/src/commands/sync.rs

```diff
@@ -354,3 +480,273 @@ async fn print_entry(doc: &Doc, entry: &Entry, content: bool) -> anyhow::Result<
+                    tracing::trace!("Progress({id}, {offset})");
+                    task_mp.add_progress(id, offset).await;
+                }
+                AddProgress::Done { hash, id } => {
+                    tracing::trace!("Done({id},{hash:?})");
+                    task_mp.adding_to_doc(id).await;
+                    match collections.get_mut(&id) {
+                        Some((path_str, size, ref mut h)) => {
+                            *h = Some(hash);
+                            let key = match key_from_path_str(
+                                root.clone(),
+                                prefix.clone(),
+                                path_str.clone(),
+                            ) {
+                                Ok(k) => k,
+                                Err(e) => {
+                                    doc_set_updates
+                                        .send(DocSetStreamUpdate::Abort)
+                                        .await
+                                        .expect("receiver dropped");
+                                    anyhow::bail!("issue creating key for entry {hash:?}: {e}");
+                                }
+                            };
+                            // send update to doc
+                            doc_set_updates
+                                .send(DocSetStreamUpdate::Entry {
+                                    id,
+                                    key,
+                                    hash,
+                                    size: *size,
+                                })
+                                .await
+                                .expect("receiver dropped");
+                        }
+                        None => {
+                            anyhow::bail!("Got Done for unknown collection id {id}");
+                        }
+                    }
+                }
+                AddProgress::AllDone { hash, .. } => {
+                    tracing::trace!("AddProgress::AllDone({hash:?})");
+                    break;
+                }
+                AddProgress::Abort(e) => {
+                    task_mp.error().await;
+                    doc_set_updates
+                        .send(DocSetStreamUpdate::Abort)
+                        .await
+                        .expect("receiver dropped");
+                    anyhow::bail!("Error while adding data: {e}");
+                }
+            }
+        }
+        Ok(())
+    });
+    let mut entries = 0;
+    let mut total_size = 0;
+    while let Some(res) = doc_set_progress.next().await {
+        match res? {
+            DocSetProgress::Done { id, size, .. } => {
+                mp.done(id).await;
+                entries += 1;
+                total_size += size;
+            }
+            DocSetProgress::AllDone => {
+                mp.all_done().await;
+                add_progress_task.abort();
+                return Ok((entries, total_size));
```

Comment:
I would prefer if we would break here instead of return. So

```rust
break Ok((entries, total_size));
```

and 

```rust
break Err(anyhow!("..."))
```

below.

Then aborting the task can happen afterwards, and at the end just return res.

Alternatively you could wrap the add_progress_task in an AbortingJoinHandle, which we got in some utils somewhere.